src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,104 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private long maxId;
+    private List<Product> products;
+    private ReadWriteLock lock;
+
+    public ArrayListProductDao() {
+        lock = new ReentrantReadWriteLock();
+        products = new ArrayList<>();
+        saveSampleProducts();
+    }
+
     @Override
     public Product getProduct(Long id) {
```

Comment:
This method should be refactored. 1) If you add throws to interface also add here. 2) When you plan to throw checked exception (like you did) than catch directive should be outside this method. In this code you catch you own exception here and return null, but outside this function exception expected cause you wrote throws ProductNotFoundException. So outside this method you implement (in general how it should be done): * if method has throws ... - try catch is implemented, * if method returns Optional<...> - isPresent is used , * if nothing from this used - !=null check is implemented. This is common error handling options in such operations.